unix/vncconnect/vncconnect.man
'\" t
.\" ** The above line should force tbl to be a preprocessor **
.\" Man page for X vncconnect
.\"
.\" Copyright (C) 2010, 2013, 2018 D. R. Commander.
.\" Copyright (C) 2005-2008 Sun Microsystems, Inc.
.\" Copyright (C) 2001-2003 Constantin Kaplinsky
.\" Copyright (C) 2000, 2001 Red Hat, Inc.
.\"
.\" You may distribute under the terms of the GNU General Public
.\" License as specified in the file LICENCE.TXT that comes with the
.\" TurboVNC distribution.
.\"
.TH vncconnect 1 "October 2018" "" "TurboVNC"
.SH NAME
vncconnect \- connect a TurboVNC session to a listening VNC viewer
.SH SYNOPSIS
.nf
\fBvncconnect\fR [\-display \fIVNC-display\fR] [\-disconnect] \fIclient\fR[:\fIport\fR]
.fi
.SH DESCRIPTION
\fBvncconnect\fR can be used to initiate a "reverse connection" between a
TurboVNC session and a VNC viewer that has been started in "listen mode."  The
viewer should be running on the machine specified by \fIclient\fR and listening
on the port specified by \fIport\fR (if \fIport\fR is not specified, then the
default of 5500 is used.)
.SH OPTIONS
.TP
\fB\-display\fR \fIVNC-display\fR
The X display of the TurboVNC session that you wish to connect to a listening
viewer (default: read from the \fBDISPLAY\fR environment variable)
.TP
\fB\-disconnect\fR
Disconnect all listening viewers from a particular TurboVNC session
.SH EXIT CODES
When used to initiate a reverse connection, \fBvncconnect\fR will exit with
status 0 if the connection was successful.  It will exit with status 1 if the
connection failed (which would be the case if there was no listening viewer
running on the specified client and port, if a firewall prevented the
connection, or if reverse connections had been disallowed in the TurboVNC
Server's security configuration file.)

When used to disconnect reverse connections, \fBvncconnect\fR will exit with
status 0 if one or more listening viewers were successfully disconnected.
Otherwise, it will exit with status 1.
.SH SEE ALSO
\fBvncviewer\fR(1), \fBvncserver\fR(1), \fBXvnc\fR(1), \fBvncpasswd\fR(1)
.SH AUTHORS
VNC was originally developed at AT&T Laboratories Cambridge. TightVNC
additions were implemented by Constantin Kaplinsky. TurboVNC, based
on TightVNC, is provided by The VirtualGL Project. Many other people
participated in development, testing and support.

\fBMan page authors:\fR
.br
Tim Waugh <twaugh@redhat.com>
.br
Constantin Kaplinsky <const@tightvnc.com>
.br
D. R. Commander <information@turbovnc.org>


ChangeLog.md
3.1
===

### Significant changes relative to 3.0.1:

1. The TurboVNC Server and Viewer now support UTF-8 clipboard transfers.

2. The TurboVNC Viewer Options dialog has been improved in the following ways:

     - The "Global" tab now has a button that can be used to reset all options
to their default values.

3. The TurboVNC Viewer now maintains a different set of options for each unique
TurboVNC host.

4. The TurboVNC Server and Viewer now implement the QEMU Extended Key Event,
QEMU LED State, and VMware LED State RFB extensions, which allow raw keyboard
scancodes to be transmitted to the VNC server instead of X11 keysyms.
Effectively, this means that the mapping of keycodes into keysyms is performed
on the host rather than the client, which eliminates various system-specific
and locale-specific key mapping issues (including issues with dead keys on
international keyboards.)  This feature also fixes lock key synchronization
issues when using the TurboVNC Viewer with VMware's VNC server.

5. If the `NoReconnect` parameter is unset (which it is by default), the
TurboVNC Viewer will now offer to reconnect if the initial connection or
authentication fails.

6. The TurboVNC Server can now listen on a Unix domain socket, rather than a
TCP port, for connections from VNC viewers.  This is useful in conjunction with
SSH tunneling, and it also allows the Unix domain socket permissions to act as
an authentication mechanism for a TurboVNC session.  Two new Xvnc arguments,
`-rfbunixpath` and `-rfbunixmode`, can be used to specify the Unix domain
socket path and, optionally, the permissions.  A new `vncserver` argument,
`-uds`, causes the script to automatically choose an appropriate Unix domain
socket path under the TurboVNC user directory.  See the `Xvnc` and `vncserver`
man pages for more details.

7. The TurboVNC Viewer can now connect to a VNC server that is listening on a
Unix domain socket.  This is accomplished by specifying `{host}::{uds_path}` as
the VNC server, where `{host}` is the hostname or IP address of the VNC host
and `{uds_path}` is the path to the Unix domain socket on the host.  If
`{host}` is not `localhost`, then SSH tunneling with an external SSH client is
implied.  Refer to the TurboVNC Viewer's usage screen (specifically, the
documentation of the `Server` parameter) and the TurboVNC User's Guide for more
details.


3.0.2
=====

### Significant changes relative to 3.0.1:

1. The Linux/Un\*x TurboVNC Viewer now works around an issue whereby Java on
Un*x platforms generates a key release event without a corresponding key press
event for dead keys on international keyboards.

2. The TurboVNC Viewer no longer pops up the F8 menu if a modifier key (Shift,
Ctrl, Alt, etc.) is pressed along with the menu key.

3. Hotkeys in the Windows TurboVNC Viewer can no longer be triggered using the
RCtrl-LAlt-Shift modifier sequence, because Windows uses RCtrl-LAlt to
represent AltGr on international keyboards.

4. The `vncserver` script now searches **/usr/local/share/fonts** for X11
fonts, which fixes an issue whereby the TurboVNC X server had few available
fonts when running on recent FreeBSD releases.

5. The Windows TurboVNC Viewer no longer overrides Java's default choice of
high DPI scaling algorithms (nearest-neighbor interpolation) when using
integral display scaling factors such as 200%.  The viewer now emulates the
behavior of Windows, using bilinear interpolation with fractional display
scaling factors (per 3.0[6]) and nearest-neighbor interpolation with integral
display scaling factors.  This improves the sharpness of text when using the
Windows TurboVNC Viewer with integral display scaling factors.  The
`turbovnc.scalingalg` system property can be set to `bicubic`, `bilinear`, or
`nearestneighbor` to override the TurboVNC Viewer's default algorithm choice.


3.0.1
=====

### Significant changes relative to 3.0:

1. Fixed an error ("JRELoadError") that occurred when attempting to use the Mac
TurboVNC Viewer app with a custom JRE based on OpenJDK 17.

2. The TurboVNC Viewer's built-in SSH client now supports the OpenSSH v1
private key format.  This fixes an "invalid privatekey" error that occurred
when attempting to use a private key generated by a recent version of OpenSSH
with the TurboVNC Viewer without `ssh-agent` or Pageant.

3. The TurboVNC Viewer's built-in SSH client now supports the `rsa-sha2-256`
and `rsa-sha2-512` signature schemes (RSA keys with, respectively, the SHA-256
and SHA-512 hash algorithms.)  This improves security and compatibility with
recent OpenSSH releases.  (OpenSSH v8.8 no longer supports the `ssh-rsa`
signature scheme by default, since `ssh-rsa` uses the now-defeated SHA-1 hash
algorithm.)  The built-in SSH client now also supports the
`PubkeyAcceptedAlgorithms` keyword in OpenSSH config files, which can be used
to disable specific signature schemes that both the client and server support.

4. The TurboVNC Server and Viewer now truncate both incoming and outgoing
clipboard updates to the number of bytes specified by the `-maxclipboard` Xvnc
argument or the `MaxClipboard` TurboVNC Viewer parameter.  Previously the
server only truncated incoming clipboard updates, and the viewer truncated
outgoing clipboard updates while ignoring incoming clipboard updates larger
than 256 kB.

5. Fixed an issue in the TurboVNC Viewer whereby specifying a key other than a
function key (`F1` through `F12`) using the `MenuKey` parameter caused "F1" to
be selected in the "Menu key" combo box in the TurboVNC Viewer Options dialog.

6. Fixed an error ("Security type not supported") that occurred when attempting
to connect to a TurboVNC session using the TurboVNC Session Manager if the
"VNC" security type was disabled in the TurboVNC Viewer (by way of the
`SecurityTypes`, `SendLocalUserName`, or `User` parameter or a corresponding
check box in the TurboVNC Viewer Options dialog) and the `SessMgrAuto`
parameter was enabled.


unix/vncconnect/vncconnect.c
/* Copyright (C) 2013, 2015, 2018 D. R. Commander.  All Rights Reserved.
 * Copyright (C) 2002-2005 RealVNC Ltd.  All Rights Reserved.
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
 * USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include "vncExt.h"


static char *programName;


static void usage(void)
{
  fprintf(stderr, "\nUSAGE: %s [-display <d>] [-disconnect] client[:port]\n\n",
          programName);
  fprintf(stderr, "-display <d> = specify the X display of the TurboVNC session that you wish to\n"
                  "               connect to a listening viewer (for instance, :1).  If this is\n"
                  "               not specified, then the value of the DISPLAY environment\n"
                  "               variable is used.\n");
  fprintf(stderr, "-id <id> = specify the ID number of the TurboVNC session, if connecting to an\n"
                  "           instance of the UltraVNC Repeater in Mode II.\n");
  fprintf(stderr, "-disconnect = disconnect all listening viewers from the specified TurboVNC\n"
                  "              session.\n\n");
  exit(1);
}


int main(int argc, char **argv)
{
  char *displayname = NULL;
  Display *dpy;  int id = -1;
  int i, disconnect = 0, status = 0;

  programName = argv[0];

  for (i = 1; i < argc; i++) {
    if (argv[i][0] != '-')
      break;

    if (!strncmp(argv[i], "-disp", 5)) {
      if (++i >= argc) usage();
      displayname = argv[i];
    } else if (!strncmp(argv[i], "-disc", 5)) {
      disconnect = 1;
    } else if (!strncmp(argv[i], "-id", 3)) {
      int _id = -1;

      if (++i >= argc) usage();
      _id = atoi(argv[i]);
      if (_id <= 0) {
        fprintf(stderr, "ERROR: ID must be greater than 0.\n");
        exit(1);
      }
      id = _id;
    } else usage();
  }

  if (argc != i + 1 && !disconnect)
    usage();

  if (!(dpy = XOpenDisplay(displayname))) {
    fprintf(stderr, "%s: unable to open display \"%s\"\n", programName,
            XDisplayName(displayname));
    exit(1);
  }

  if (disconnect) {
    if (!XVncExtConnect(dpy, "")) {
      fprintf(stderr, "Could not disconnect listening viewers (perhaps there are none)\n");
      status = 1;
    }
  } else {
    char temps[256];

    if (id >= 0) {
      strncpy(temps, argv[i], 255);
      temps[255] = 0;
      snprintf(&temps[strlen(temps)], 255 - strlen(temps), "#%d", id);
    }
    if (!XVncExtConnect(dpy, id >= 0 ? temps : argv[i])) {
      fprintf(stderr, "Reverse connection to %s failed\n", argv[i]);
      status = 1;
    }
  }

  XCloseDisplay(dpy);

  return status;
}
